### ISSUE TITLE
Build failure: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]

### ISSUE THREAD

**Issue by mikrosk** [dxx-rebirth/dxx-rebirth#777]
Using the same environment and source tree as in #776 plus a cross-compiler (m68k-atari-mintelf-g++ 13.2) I've got the same bugs and additionally one more warning:

```
m68k-atari-mintelf-g++ -c -g -O2 -ftabstop=4 -Wall -Werror=extra -Werror=format=2 -Werror=missing-braces -Werror=missing-include-dirs -Werror=uninitialized -Werror=undef -Werror=pointer-arith -Werror=cast-qual -Werror=missing-declarations -Werror=vla -m68020-60 -fomit-frame-pointer -D__STDC_LIMIT_MACROS -Wno-error=format-truncation -funsigned-char -std=gnu++20 -Werror=unused -Werror=useless-cast -Wimplicit-fallthrough=5 -fstrict-flex-arrays -fvisibility=hidden -Wduplicated-branches -Wduplicated-cond -Wstrict-flex-arrays -Wsuggest-attribute=noreturn -Wsuggest-final-types -Wsuggest-override -Wlogical-op -Wold-style-cast -Wredundant-decls -Wno-sign-compare -DPHYSFS_DEPRECATED= -DDXX_USE_SHAREPATH=1 -DNDEBUG -DRELEASE -Icommon/include -Icommon/main -I. -Ibuild -I/home/mikro/gnu-tools/m68000/m68k-atari-mintelf/sys-root/usr/include -I/home/mikro/gnu-tools/m68000/m68k-atari-mintelf/sys-root/usr/include/SDL common/arch/sdl/timer.cpp -o build/common/arch/sdl/timer.o
In file included from common/include/dsx-ns.h:9,
                 from common/include/maths.h:21,
                 from common/arch/sdl/timer.cpp:15:
common/include/physfsx.h: In function 'PHYSFS_sint64 dcx::PHYSFSX_check_readBytes(PHYSFS_File*, V*, PHYSFS_uint64)':
common/include/physfsx.h:69:44: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:53: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                     ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:44: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:61: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                             ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h: In function 'PHYSFS_sint64 dcx::PHYSFSX_check_readBytes(PHYSFS_File*, std::array<_Tp, _Nm>&, PHYSFS_uint64)':
common/include/physfsx.h:79:35: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   79 |     (void)(_DXX_PHYSFS_CHECK_SIZE(len, sizeof(V) * std::size(buffer)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds array size"), 0));
      |                                   ^~~
build/dxxsconf.h:39:53: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                     ^
common/include/physfsx.h:79:12: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   79 |     (void)(_DXX_PHYSFS_CHECK_SIZE(len, sizeof(V) * std::size(buffer)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds array size"), 0));
      |            ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:79:35: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   79 |     (void)(_DXX_PHYSFS_CHECK_SIZE(len, sizeof(V) * std::size(buffer)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds array size"), 0));
      |                                   ^~~
build/dxxsconf.h:39:61: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                             ^
common/include/physfsx.h:79:12: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   79 |     (void)(_DXX_PHYSFS_CHECK_SIZE(len, sizeof(V) * std::size(buffer)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds array size"), 0));
      |            ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h: In function 'PHYSFS_sint64 dcx::PHYSFSX_check_writeBytes(PHYSFS_File*, const V*, PHYSFS_uint64)':
common/include/physfsx.h:96:45: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   96 |     DXX_PHYSFS_CHECK_WRITE_SIZE_OBJECT_SIZE(len, buffer);
      |                                             ^~~
build/dxxsconf.h:39:53: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                     ^
common/include/physfsx.h:52:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   52 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("write size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:96:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_WRITE_SIZE_OBJECT_SIZE'
   96 |     DXX_PHYSFS_CHECK_WRITE_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:96:45: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   96 |     DXX_PHYSFS_CHECK_WRITE_SIZE_OBJECT_SIZE(len, buffer);
      |                                             ^~~
build/dxxsconf.h:39:61: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                             ^
common/include/physfsx.h:52:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   52 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("write size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:96:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_WRITE_SIZE_OBJECT_SIZE'
   96 |     DXX_PHYSFS_CHECK_WRITE_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h: In instantiation of 'PHYSFS_sint64 dcx::PHYSFSX_check_readBytes(PHYSFS_File*, V*, PHYSFS_uint64) [with V = unsigned char; PHYSFS_sint64 = long long int; PHYSFS_File = PHYSFS_File; PHYSFS_uint64 = long long unsigned int]':
common/include/physfsx.h:157:6:   required from here
common/include/physfsx.h:69:44: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:53: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                     ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:44: warning: narrowing conversion of '(PHYSFS_uint64)len' from 'PHYSFS_uint64' {aka 'long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:53: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                     ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:44: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:61: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                             ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:44: warning: narrowing conversion of '(PHYSFS_uint64)len' from 'PHYSFS_uint64' {aka 'long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:61: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                             ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h: In instantiation of 'PHYSFS_sint64 dcx::PHYSFSX_check_readBytes(PHYSFS_File*, V*, PHYSFS_uint64) [with V = signed char; PHYSFS_sint64 = long long int; PHYSFS_File = PHYSFS_File; PHYSFS_uint64 = long long unsigned int]':
common/include/physfsx.h:388:9:   required from here
common/include/physfsx.h:69:44: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:53: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                     ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:44: warning: narrowing conversion of '(PHYSFS_uint64)len' from 'PHYSFS_uint64' {aka 'long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:53: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                     ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:44: warning: narrowing conversion of 'len' from 'const PHYSFS_uint64' {aka 'const long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:61: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                             ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:44: warning: narrowing conversion of '(PHYSFS_uint64)len' from 'PHYSFS_uint64' {aka 'long long unsigned int'} to 'std::size_t' {aka 'long unsigned int'} [-Wnarrowing]
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |                                            ^~~
build/dxxsconf.h:39:61: note: in definition of macro 'DXX_CONSTANT_TRUE'
   39 | #define DXX_CONSTANT_TRUE(E) (__builtin_constant_p((E)) && (E))
      |                                                             ^
common/include/physfsx.h:50:79: note: in expansion of macro '_DXX_PHYSFS_CHECK_SIZE'
   50 |     (void)(__builtin_object_size(BUFFER_PTR, 1) != static_cast<size_t>(-1) && _DXX_PHYSFS_CHECK_SIZE(ELEMENT_SIZE, __builtin_object_size(BUFFER_PTR, 1)) && (DXX_ALWAYS_ERROR_FUNCTION("read size exceeds element size"), 0))
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~
common/include/physfsx.h:69:5: note: in expansion of macro 'DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE'
   69 |     DXX_PHYSFS_CHECK_READ_SIZE_OBJECT_SIZE(len, buffer);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This repeats for other files too, generating a lot of clutter. The fix is easy (https://github.com/mikrosk/dxx-rebirth/commit/68161355c21ab20d9c05b1c7deb418ad4a586851) but maybe there's a cleaner solution. `std::size_t` is 32-bit  on 32-bit systems so it can't be used as a replacement for 64-bit integers from PHYSFS.

[dxxsconf.h.zip](https://github.com/dxx-rebirth/dxx-rebirth/files/15225587/dxxsconf.h.zip)

**Comment by vLKp** [comment]
Errors in macros tend to be noisy, and errors in macros used in header files are even worse.

Yes, switching to a guaranteed 64-bit type is a reasonable solution.

**Comment by mikrosk** [comment]
Fixed by 0e90af1c.
